amd_tressfx_vulkan/src/TressFXSimulationVulkan.cpp

```diff
@@ -0,0 +1,1180 @@
+                                                  m_pTressFXMesh->m_UpdateFollowHairSet};
+        vkCmdBindDescriptorSets(
+            commandBuffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+            m_CSUpdateFollowHairVerticesLayout, 0, AMD_ARRAY_SIZE(updateFollowHairSets),
+            updateFollowHairSets, AMD_ARRAY_SIZE(descriptorOffset), descriptorOffset);
+        vkCmdBindPipeline(commandBuffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+                          m_CSUpdateFollowHairVertices);
+        vkCmdDispatch(commandBuffer, numOfGroupsForCS_VertexLevel, 1, 1);
+
+        VkBufferMemoryBarrier updateFollowHairBarriers[] = {
+            getBufferBarrier(m_pTressFXMesh->m_HairVertexPositionsBuffer,
+                             VK_ACCESS_SHADER_WRITE_BIT | VK_ACCESS_SHADER_READ_BIT,
+                             VK_ACCESS_SHADER_READ_BIT),
+        };
+
+        vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT,
+                             VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, 0, 0, nullptr,
+                             AMD_ARRAY_SIZE(updateFollowHairBarriers),
+                             updateFollowHairBarriers, 0, nullptr);
+    }
+
+    VkBufferMemoryBarrier makeTangentBarrier[] = {getBufferBarrier(
+        m_pTressFXMesh->m_HairVertexTangentsBuffer, VK_ACCESS_MEMORY_READ_BIT,
+        VK_ACCESS_SHADER_READ_BIT | VK_ACCESS_MEMORY_WRITE_BIT)};
+
+    vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT,
+                         VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT, 0, 0, nullptr,
+                         AMD_ARRAY_SIZE(makeTangentBarrier), makeTangentBarrier, 0,
+                         nullptr);
+
+    // Compute tangents for every vertex (guide + follow)
+    {
+        int numOfGroupsForCS_TotalVertexLevel =
+            (int)(((float)(m_pTressFXMesh->m_HairAsset.m_NumTotalHairVertices) /
+                   (float)THREAD_GROUP_SIZE) *
+                  density);
+        vkCmdBindPipeline(commandBuffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+                          m_CSComputeTangents);
+        VkDescriptorSet computeTangentSet[] = {m_configSet,
+                                               m_pTressFXMesh->m_ComputeTangentSet};
+        vkCmdBindDescriptorSets(commandBuffer, VK_PIPELINE_BIND_POINT_COMPUTE,
+                                m_CSComputeTangentsLayout, 0,
+                                AMD_ARRAY_SIZE(computeTangentSet), computeTangentSet,
+                                AMD_ARRAY_SIZE(descriptorOffset), descriptorOffset);
+        vkCmdDispatch(commandBuffer, numOfGroupsForCS_TotalVertexLevel, 1, 1);
+
+        VkBufferMemoryBarrier computeTangentBarrier[] = {
+            getBufferBarrier(m_pTressFXMesh->m_HairVertexTangentsBuffer,
+                             VK_ACCESS_MEMORY_READ_BIT | VK_ACCESS_MEMORY_WRITE_BIT,
+                             VK_ACCESS_SHADER_READ_BIT)};
+
+        vkCmdPipelineBarrier(commandBuffer, VK_PIPELINE_STAGE_COMPUTE_SHADER_BIT,
```

Comment:
The barriers here should be batched up. As far as I can tell, they affect different resources, so there's not need to actually barrier multiple times. Move them all to the end and execute in one go.